### ISSUE TITLE
bug in crossLevel sampler that causes incorrect behavior for non-identity links

### ISSUE THREAD

**Issue by paciorek** [nimble-dev/nimble#924]
There appears to be a long-standing bug in the cross-level sampler (it was there in 0.6-10, possibly as long ago as when the cross-level sampler was written, unless we didn't have non-identity links at the point that it was written). The bug causes incorrect behavior for non-identity link functions. I believe that the resulting sampler behavior would generally be so egregiously bad that a user would not make use of the results. Hopefully that is the case.

The problem is in the run function of the crossLevel sampler, we call `lowConjugateGetLogDensityFunctions[[iSF]]$run`, which calls `conjugateSamplerFunction$getPosteriorLogDensity()`. In that function, we set `model[[target]]`to 1 and never set it back to its original value (see code below for an example with normal-normal conjugacy). As a result, the resulting posterior density that is used is incorrect (so the acceptance calculation is invalid) and the value of the dependent nodes (those nodes that are dependents of the top-level nodes that are 'officially' the ones being sampled) are set to 1. Note that in the step of actually sampling from the conjugate distribution, we sample reasonable values, but then there is another pass through `conjugateSamplerFunction$getPosteriorLogDensity()`  and the dependent nodes are set to 1.  So once a proposal is accepted, all the dependent nodes will be 1, and the MCMC goes haywire.

Also @danielturek - a side question -- crossLevel sampler is written for a block of (hyper)parameters. I presume it is fine for a scalar hyperparameter, even though the RW proposals for the (hyper)parameters are implicitly set up for a block of parameters?

```
Class method definition for method getPosteriorLogDensity()
function () 
{
    prior_mean <- model$getParam(target[1], "mean")
    prior_tau <- model$getParam(target[1], "tau")
    for (iDep in 1:N_dep_dnorm) {
        dep_dnorm_values[iDep] <<- model$getParam(dep_dnorm_nodeNames[iDep], 
            "value")
        dep_dnorm_tau[iDep] <<- model$getParam(dep_dnorm_nodeNames[iDep], 
            "tau")
    }
    model[[target]] <<- 0
    model$calculate(calcNodesDeterm)
    for (iDep in 1:N_dep_dnorm) dep_dnorm_offset[iDep] <<- model$getParam(dep_dnorm_nodeNames[iDep], 
        "mean")
    model[[target]] <<- 1
    model$calculate(calcNodesDeterm)
    for (iDep in 1:N_dep_dnorm) dep_dnorm_coeff[iDep] <<- model$getParam(dep_dnorm_nodeNames[iDep], 
        "mean") - dep_dnorm_offset[iDep]
    contribution_mean <<- 0
    contribution_tau <<- 0
    for (iDep in 1:N_dep_dnorm) {
        contribution_mean <<- contribution_mean + dep_dnorm_coeff[iDep] * 
            (dep_dnorm_values[iDep] - dep_dnorm_offset[iDep]) * 
            dep_dnorm_tau[iDep]
        contribution_tau <<- contribution_tau + dep_dnorm_coeff[iDep]^2 * 
            dep_dnorm_tau[iDep]
    }
    targetValue <- model[[target]]
    posteriorLogDensity <- dnorm(targetValue, mean = (prior_mean * 
        prior_tau + contribution_mean)/(prior_tau + contribution_tau), 
        sd = (prior_tau + contribution_tau)^(-0.5), log = 1)
    return(posteriorLogDensity)
}
```

**Comment by danielturek** [comment]
Hmm....  How is it that this has gone unnoticed for so long?  I feel like this is the original implementation of the crossLevel sampler, the implementation of which required extending the system for conjugate samplers to include the `getPosteriorLogDensity` functions, but apparently the way those were implemented modifies the target node value, thus making other, later, calculations erroneous -- does that sound about right?

So, if I'm reading correctly, it sounds like the `getPosteriorLogDensity` of each conjugate sampler would need to be modified, to make sure the target node is restored to its proper original value, e.g. including code along the lines of:
```
model[[target]] <<- originalTargetNodeValue
```

Does that sound about right?

**Comment by danielturek** [comment]
@paciorek See PR #925, and specifically the branch `crossLevelFix`.  This was quick work, but might be the correct (and reasonably straightforward) fix to the problem you described.

I opened the above PR #925 to trigger testing of the conjugate samplers.

When you have time, could you please build off this branch (`crossLevelFix`), and take a look at the new conjugate sampler `getPosteriorLogDensity` method code, and assess whether it fixes the problem you observed?

Thanks.

**Comment by danielturek** [comment]
@paciorek Thoughts about this fix?

**Comment by paciorek** [comment]
closed by PR #925.
